File: verify/go/verify_one.go
// Command verify_one re-verifies the committed Phase 2 evidence trail offline
// with the reference Go SDK — the SAME committed bytes the TypeScript
// scripts/verify-one.ts consumes. It decodes each ProofBundle and its
// VerificationReceipt, then checks, per claim:
//
//   (1) bundle  -> identity_status reproduces the recorded decision
//   (2) receipt -> signature verifies
//   (3) receipt -> bundle_hash binding matches BundleHash(bundle)
//   (4) receipt -> prev_hash chain link is intact (genesis = 32 zero bytes)
//
// Usage (from verify/go/):
//
//	go run .            # verify every claim in manifest order (checks the prev_hash chain)
//	go run . commit-1   # verify a single claim by label
//
// Exits non-zero if any check fails. Offline; no network.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	ratify "github.com/identities-ai/ratify-protocol"
)

type manifestClaim struct {
	Label            string `json:"label"`
	RequestedPath    string `json:"requested_path"`
	VerifiedAt       int64  `json:"verified_at"`
	ExpectedDecision string `json:"expected_decision"`
	Revoked          bool   `json:"revoked"`
}

type manifestFile struct {
	TargetResourceID string          `json:"target_resource_id"`
	RevokedCerts     []string        `json:"revoked_certs"`
	Claims           []manifestClaim `json:"claims"`
}

// setRevocationProvider is the Go analogue of the harness SetRevocationProvider:
// an in-memory revocation set (SPEC §17.1).
type setRevocationProvider struct{ revoked map[string]bool }

func (p setRevocationProvider) IsRevoked(certID string) (bool, error) {
	return p.revoked[certID], nil
}

func main() {
	// evidence/ sits two levels up from verify/go/.
	evidenceDir := filepath.Join("..", "..", "evidence")

	raw, err := os.ReadFile(filepath.Join(evidenceDir, "manifest.json"))
	if err != nil {
		fatal("reading manifest: %v", err)
	}
	var manifest manifestFile
	if err := json.Unmarshal(raw, &manifest); err != nil {
		fatal("parsing manifest: %v", err)
	}

	revoked := make(map[string]bool, len(manifest.RevokedCerts))
	for _, c := range manifest.RevokedCerts {
		revoked[c] = true
	}
	revocation := setRevocationProvider{revoked: revoked}

	// Optional single-label filter. Note: verifying one claim in isolation
	// still checks the prev_hash link against the preceding receipt in the
	// full chain, so we always walk the chain in order and only *report* the
	// requested label.
	onlyLabel := ""
	if len(os.Args) > 1 {
		onlyLabel = os.Args[1]
	}

	var prevReceiptHash []byte
	zero32 := make([]byte, 32)
	failures := 0
	reported := 0

	for _, entry := range manifest.Claims {
		bundleBytes, err := os.ReadFile(filepath.Join(evidenceDir, "bundles", entry.Label+".json"))
		if err != nil {
			fatal("reading bundle %s: %v", entry.Label, err)
		}
		receiptBytes, err := os.ReadFile(filepath.Join(evidenceDir, "receipts", entry.Label+".json"))
		if err != nil {
			fatal("reading receipt %s: %v", entry.Label, err)
		}

		bundle, err := ratify.DecodeProofBundle(bundleBytes)
		if err != nil {
			fatal("decoding bundle %s: %v", entry.Label, err)
		}
		receipt, err := ratify.DecodeVerificationReceipt(receiptBytes)
		if err != nil {
			fatal("decoding receipt %s: %v", entry.Label, err)
		}

		// (1) Re-verify the bundle; identity_status must reproduce the recorded decision.
		opts := ratify.VerifyOptions{
			RequiredScope: ratify.ScopeFilesWrite,
			Now:           time.Unix(entry.VerifiedAt, 0),
			Context: ratify.VerifierContext{
				RequestedResourceID: manifest.TargetResourceID,
				RequestedPath:       entry.RequestedPath,
				HasResource:         true,
			},
		}
		if entry.Revoked {
			opts.Revocation = revocation
		}
		res := ratify.Verify(bundle, opts)
		bundleOK := res.IdentityStatus == entry.ExpectedDecision

		// (2) Receipt signature.
		sigErr := ratify.VerifyVerificationReceipt(receipt)

		// (3) bundle_hash binding.
		bh, err := ratify.BundleHash(bundle)
		if err != nil {
			fatal("hashing bundle %s: %v", entry.Label, err)
		}
		hashOK := bytes.Equal(receipt.BundleHash, bh)

		// (4) prev_hash chain link.
		expectedPrev := prevReceiptHash
		if expectedPrev == nil {
			expectedPrev = zero32
		}
		chainOK := bytes.Equal(receipt.PrevHash, expectedPrev)

		// Advance the chain regardless of the label filter so links stay honest.
		rh, err := ratify.ReceiptHash(receipt)
		if err != nil {
			fatal("hashing receipt %s: %v", entry.Label, err)
		}
		prevReceiptHash = rh

		if onlyLabel != "" && entry.Label != onlyLabel {
			continue
		}
		reported++

		receiptOK := sigErr == nil && hashOK && chainOK
		claimOK := bundleOK && receiptOK
		if !claimOK {
			failures++
		}

		sig := "ok"
		if sigErr != nil {
			sig = fmt.Sprintf("BAD (%v)", sigErr)
		}
		fmt.Printf("%s: identity_status=%s (recorded %s) %s\n",
			entry.Label, res.IdentityStatus, entry.ExpectedDecision, okmiss(bundleOK, "OK", "MISMATCH"))
		fmt.Printf("%s: receipt signature=%s bundle_hash_binding=%s prev_hash_chain=%s %s\n",
			entry.Label, sig, okmiss(hashOK, "ok", "BAD"), okmiss(chainOK, "ok", "BROKEN"), okmiss(receiptOK, "OK", "FAIL"))
	}

	if onlyLabel != "" && reported == 0 {
		fatal("no claim labelled %q in manifest", onlyLabel)
	}
	if failures > 0 {
		fmt.Fprintf(os.Stderr, "verify_one: %d claim(s) FAILED\n", failures)
		os.Exit(1)
	}
}

func okmiss(ok bool, yes, no string) string {
	if ok {
		return yes
	}
	return no
}

func fatal(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "verify_one: FAILED — "+format+"\n", args...)
	os.Exit(1)
}

File: verify/python/verify_one.py
#!/usr/bin/env python3
"""Re-verify the committed Phase 2 evidence trail offline with the Python SDK.

Consumes the SAME committed bytes as scripts/verify-one.ts (TypeScript) and
verify/go/verify_one.go (Go). For each claim it decodes the ProofBundle and its
VerificationReceipt, then checks:

  (1) bundle  -> identity_status reproduces the recorded decision
  (2) receipt -> signature verifies
  (3) receipt -> bundle_hash binding matches bundle_hash(bundle)
  (4) receipt -> prev_hash chain link is intact (genesis = 32 zero bytes)

Usage (from verify/python/, with the local SDK installed — see VERIFY.md):

    python3 verify_one.py            # verify every claim in manifest order
    python3 verify_one.py commit-1   # verify a single claim by label

Exits non-zero if any check fails. Offline; no network.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

from ratify_protocol import (
    SCOPE_FILES_WRITE,
    VerifierContext,
    VerifyOptions,
    bundle_hash,
    decode_proof_bundle,
    decode_verification_receipt,
    receipt_hash,
    verify_bundle,
    verify_verification_receipt,
)

# evidence/ sits two levels up from verify/python/.
EVIDENCE_DIR = Path(__file__).resolve().parent.parent.parent / "evidence"


class SetRevocationProvider:
    """In-memory revocation set (SPEC §17.1); Python analogue of the harness provider."""

    def __init__(self, revoked: set[str]) -> None:
        self._revoked = revoked

    def is_revoked(self, cert_id: str):
        return (cert_id in self._revoked, None)


def _okmiss(ok: bool, yes: str, no: str) -> str:
    return yes if ok else no


def main() -> int:
    manifest = json.loads((EVIDENCE_DIR / "manifest.json").read_text())
    target_resource_id = manifest["target_resource_id"]
    revocation = SetRevocationProvider(set(manifest["revoked_certs"]))

    only_label = sys.argv[1] if len(sys.argv) > 1 else ""

    prev_receipt_hash: bytes | None = None
    zero32 = bytes(32)
    failures = 0
    reported = 0

    for entry in manifest["claims"]:
        label = entry["label"]
        bundle = decode_proof_bundle((EVIDENCE_DIR / "bundles" / f"{label}.json").read_text())
        receipt = decode_verification_receipt((EVIDENCE_DIR / "receipts" / f"{label}.json").read_text())

        # (1) Re-verify the bundle; identity_status must reproduce the recorded decision.
        opts = VerifyOptions(
            required_scope=SCOPE_FILES_WRITE,
            now=entry["verified_at"],
            context=VerifierContext(
                requested_resource_id=target_resource_id,
                requested_path=entry["requested_path"],
                has_resource=True,
            ),
            revocation=revocation if entry["revoked"] else None,
        )
        res = verify_bundle(bundle, opts)
        bundle_ok = res.identity_status == entry["expected_decision"]

        # (2) Receipt signature.
        sig_err = verify_verification_receipt(receipt)

        # (3) bundle_hash binding.
        hash_ok = receipt.bundle_hash == bundle_hash(bundle)

        # (4) prev_hash chain link.
        expected_prev = prev_receipt_hash if prev_receipt_hash is not None else zero32
        chain_ok = receipt.prev_hash == expected_prev

        # Advance the chain regardless of the label filter so links stay honest.
        prev_receipt_hash = receipt_hash(receipt)

        if only_label and label != only_label:
            continue
        reported += 1

        receipt_ok = sig_err is None and hash_ok and chain_ok
        claim_ok = bundle_ok and receipt_ok
        if not claim_ok:
            failures += 1

        sig = "ok" if sig_err is None else f"BAD ({sig_err})"
        print(f"{label}: identity_status={res.identity_status} "
              f"(recorded {entry['expected_decision']}) {_okmiss(bundle_ok, 'OK', 'MISMATCH')}")
        print(f"{label}: receipt signature={sig} "
              f"bundle_hash_binding={_okmiss(hash_ok, 'ok', 'BAD')} "
              f"prev_hash_chain={_okmiss(chain_ok, 'ok', 'BROKEN')} "
              f"{_okmiss(receipt_ok, 'OK', 'FAIL')}")

    if only_label and reported == 0:
        print(f"verify_one: no claim labelled {only_label!r} in manifest", file=sys.stderr)
        return 2
    if failures:
        print(f"verify_one: {failures} claim(s) FAILED", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
